### PR TITLE
[merged] Add --fstrim to atomic ps command

### PR DESCRIPTION
**Pull request by shishir-a412ed**
Signed-off-by: Shishir Mahajan <shishir.mahajan@redhat.com>

### CODE REVIEW

**Comment by shishir-a412ed** [comment]
ping @rhvgoyal @rhatdan 

/cc @giuseppe 

Shishir

File: Atomic/ps.py [hunk 1]
```diff
@@ -83,6 +96,17 @@ def ps_tty(self):
         if not all_containers:
             return
 
+        if self.args.fstrim:
+            for info in all_containers:
+                container = info["container"]
+                ret = self._inspect_container(name=container)
```
Comment (giuseppe) [inline Atomic/ps.py]: _inspect_container works only for Docker containers.  So this loop should skip any container where `info["type"] != "docker"`

File: Atomic/ps.py [hunk 2]
```diff
@@ -24,6 +30,8 @@ def cli(subparser):
                      help=_("Don't truncate output"))
     pss.add_argument("-q", "--quiet", action='store_true', dest="quiet", default=False,
                      help=_("Only display container IDs"))
+    pss.add_argument("--fstrim", action='store_true', dest="fstrim", default=False,
```
Comment (giuseppe) [inline Atomic/ps.py]: this should be added to the documentation and bash autocompletion as well.  There is a `test_ps.sh` file that you could use to add a test as well.

File: Atomic/ps.py [hunk 3]
```diff
@@ -33,6 +41,11 @@ def ps_tty(self):
         skull = (u"\u2620").encode('utf-8')
         all_container_info = self.ps()
         all_containers = []
+
+        if self.args.all and self.args.fstrim:
+            util.write_out("fstrim can only be used for running containers. Use --fstrim without -a/--all option.")
```
Comment (rhatdan) [inline Atomic/ps.py]: Can we make these mutually exclusive options in the argparser?

**Comment by rhatdan** [comment]
I am not crazy about this being an option on atomic ps.  Might be better to add a new category.  

```
atomic containers trim
```

And then have something like
```
atoimc containers ps 
```
To mimic atomic images.

**Comment by rhatdan** [comment]
Or better yet

```
atomic containers list
```

**Comment by shishir-a412ed** [comment]
@rhatdan so you are saying remove `atomic ps` completely.
And we should have something like:

1) `atomic containers list` which would have all the options like `--all`, `--no-trunc`, `--quiet` which currently `atomic ps` has. 
2) `atomic containers fstrim` which would discard unused blocks on the container rootfs.

Shishir

File: Atomic/ps.py (hunk 1 above)
Comment (shishir-a412ed) [inline Atomic/ps.py]: Fixed.

**Comment by rhatdan** [comment]
We could maybe leave `atomic ps` as a deprecated interface and make it link to `atomic containers list`

@cgwalters @baude What do you think?

Adding additional commands to atomic ps that has nothing to do with listing processes makes no sense.

**Comment by giuseppe** [comment]
we didn't keep it around for long enough that we can probably just rename it and not have to worry of a deprecated interface

**Comment by rhatdan** [comment]
Ok @shishir-a412ed Please change pull request to move to atomic containers trim,  Then we can get a different pull request to change ps to atomic containers list.

**Comment by shishir-a412ed** [comment]
@rhatdan Since `atomic containers trim` needs to list containers (ID's) and `fstrim` their rootfs one by one. It would be better to open PR for `atomic containers list` first. And once that is merged.
We can add the subcommand `trim` to `atomic containers`. 

Based on @giuseppe comment, no need to deprecate `ps` right? Just create (rename ps) a new command `atomic containers list`.

**Comment by rhatdan** [comment]
Yes I am preparing an `atomic containers list` pull request.  Just waiting for my previous pull request to get merged.
https://github.com/projectatomic/atomic/pull/570

**Comment by rhatdan** [comment]
@shishir-a412ed My atomic containers list patch has been merged, so lets get this one rebased on top of it.
`atomic containers trim`
@rhvgoyal WDYT

**Comment by rhvgoyal** [comment]
@rhatdan `atomic containers trim` sounds good to me. It is simple and intuitive.

**Comment by shishir-a412ed** [comment]
@giuseppe do we need `atomic containers trim` for both type of containers {system containers, docker containers} ?

**Comment by giuseppe** [comment]
@shishir-a412ed no, we don't need fstrim for system containers

**Comment by shishir-a412ed** [comment]
@giuseppe thanks for the quick response !

Shishir

**Comment by shishir-a412ed** [comment]
@rhatdan @giuseppe PTAL.

Shishir

**Comment by giuseppe** [comment]
LGTM

**Comment by rhvgoyal** [comment]
What is "system container" and why don't we need trimming capabilities for those. Is it because of the fact that system containers run outside of docker and they don't use thin volume?

**Comment by giuseppe** [comment]
@rhvgoyal exactly they don't use thin volumes.  Images are stored in OSTree and containers are exploded to the file system

File: docs/atomic-containers.1.md [hunk 4]
```diff
@@ -29,10 +35,20 @@ system.
 
 Using --all will list all the installed containers.
 
+**atomic containers trim**, By default this will only discard unused blocks (fstrim)
```
Comment (rhatdan) [inline docs/atomic-containers.1.md]: Discard unused blocks (fstrim) on all running containers.

File: docs/atomic-containers.1.md [hunk 5]
```diff
@@ -14,8 +14,14 @@ atomic containers allows the user to view and operate on containers
...
 
-# list OPTIONS
+**trim**
+
+discard unused blocks (fstrim) on running containers.
+
+# OPTIONS
 [**-h|--help**]
+
+# list OPTIONS
```
Comment (rhatdan) [inline docs/atomic-containers.1.md]: Discard this section, only use the list options section below.

File: Atomic/containers.py [hunk 6]
```diff
@@ -29,10 +37,25 @@ def cli(subparser):
...
+    # atomic containers trim
+    trimp = containers_subparser.add_parser("trim",
+                                            help=_("discard unused blocks (fstrim) on running containers"),
+                                            epilog="By default this will only discard unused blocks (fstrim) on rootfs of running containers.")
+    trimp.set_defaults(_class=Containers, func='ps_fstrim')
+    util.add_opt(trimp)
 
 class Containers(Atomic):
 
+    def ps_fstrim(self):
```
Comment (rhatdan) [inline Atomic/containers.py]: Lets drop the ps_ and just call this fstrim

File: Atomic/containers.py [hunk 7]
```diff
@@ -29,10 +37,25 @@ def cli(subparser):
     pss.add_argument("-q", "--quiet", action='store_true', dest="quiet", default=False,
                      help=_("Only display container IDs"))
     util.add_opt(pss)
-
+    # atomic containers trim
+    trimp = containers_subparser.add_parser("trim",
+                                            help=_("discard unused blocks (fstrim) on running containers"),
+                                            epilog="By default this will only discard unused blocks (fstrim) on rootfs of running containers.")
+    trimp.set_defaults(_class=Containers, func='ps_fstrim')
+    util.add_opt(trimp)
```
Comment (rhatdan) [inline Atomic/containers.py]: Don't want this

**Comment by shishir-a412ed** [comment]
@rhatdan PTAL.

**Comment by rhatdan** [comment]
@rh-atomic-bot r+

**Pull request closed without merging**